x/fees/keeper/fees.go

```diff
@@ -0,0 +1,189 @@
+package keeper
+
+import (
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/tharsis/evmos/v3/x/fees/types"
+)
+
+// GetAllFees - get all registered DevFeeInfo instances
+func (k Keeper) GetAllFees(ctx sdk.Context) []types.DevFeeInfo {
+	feeInfos := []types.DevFeeInfo{}
+
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.KeyPrefixFee)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		contractAddress := common.BytesToAddress(iterator.Key())
+		deployerAddress := sdk.AccAddress(iterator.Value())
+		withdrawalAddress, _ := k.GetWithdrawal(ctx, contractAddress)
+		feeInfo := types.DevFeeInfo{
+			ContractAddress: contractAddress.String(),
+			DeployerAddress: deployerAddress.String(),
+			WithdrawAddress: withdrawalAddress.String(),
+		}
+		feeInfos = append(feeInfos, feeInfo)
+	}
+
+	return feeInfos
+}
+
+// IterateFees iterates over all registered `DevFeeInfos` and performs a
+// callback.
+func (k Keeper) IterateFees(
+	ctx sdk.Context,
+	handlerFn func(fee types.DevFeeInfo) (stop bool),
+) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.KeyPrefixFee)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		contractAddress := common.BytesToAddress(iterator.Key())
+		deployerAddress := sdk.AccAddress(iterator.Value())
+		withdrawalAddress, _ := k.GetWithdrawal(ctx, contractAddress)
+		feeInfo := types.DevFeeInfo{
+			ContractAddress: contractAddress.String(),
+			DeployerAddress: deployerAddress.String(),
+			WithdrawAddress: withdrawalAddress.String(),
+		}
+
+		if handlerFn(feeInfo) {
+			break
+		}
+	}
+}
+
+// GetFeeInfo returns DevFeeInfo for a registered contract
+func (k Keeper) GetFeeInfo(ctx sdk.Context, contract common.Address) (types.DevFeeInfo, bool) {
+	deployerAddress, found := k.GetDeployer(ctx, contract)
+	if !found {
+		return types.DevFeeInfo{}, false
+	}
+	withdrawalAddress, _ := k.GetWithdrawal(ctx, contract)
+	feeInfo := types.DevFeeInfo{
+		ContractAddress: contract.String(),
+		DeployerAddress: deployerAddress.String(),
+		WithdrawAddress: withdrawalAddress.String(),
```

Comment:
same

Reply:
https://github.com/tharsis/evmos/pull/436#discussion_r842606325. The withdrawal address will not be empty if it does find a deployer at line 62.

Reply:
I actually have a test for this situation: https://github.com/tharsis/evmos/pull/436/files#diff-018d1483e38c4d82f6df127ae75758c2b870f7fc3ab5a657f66d1e5945ee0f56R62-R72, with the check here: https://github.com/tharsis/evmos/pull/436/files#diff-018d1483e38c4d82f6df127ae75758c2b870f7fc3ab5a657f66d1e5945ee0f56R139

Reply:
I see an issue with the state bloat being ignored on Export/ImportGenesis:

- `ExportGenesis` will get all the `DevFeeInfo`s
   - All of them will contain the `DeployerAddress` as well as the `WithdrawAddress`, because `GetWithdrawal` will return the deployer if there's no `WithdrawAddress`. So no `DevFeeInfo` will actually have the `WithdrawAddress` set to `""` 
- On `ImportGenesis` will set the Fee (contract -> deployer), `ReverseFee` as well as the `WithdrawAddress` (because the latter will be set to the `DeployerAddress`).

Reply:
Correct. I made the suggested changes in https://github.com/tharsis/evmos/pull/436/commits/404b219ca3740a5fb46595e330fc9817ceff2584.
Now the withdrawal address is left empty if it does not exist, including in the `grpc_query` functions.